llvm/unittests/ProfileData/InstrProfTest.cpp

```diff
@@ -64,13 +64,130 @@ struct InstrProfTest : ::testing::Test {
+    // and MutableArrayRef takes vector::data() when initialized with a vector.
+    Record.addValueData(ValueKind, i,
+                        ValueDataArray[i].empty() ? nullptr
+                                                  : ValueDataArray[i].data(),
+                        ValueDataArray[i].size(), nullptr);
+  }
+}
+
 struct SparseInstrProfTest : public InstrProfTest {
   void SetUp() override { Writer.setOutputSparse(true); }
 };
 
 struct MaybeSparseInstrProfTest : public InstrProfTest,
                                   public ::testing::WithParamInterface<bool> {
   void SetUp() override { Writer.setOutputSparse(GetParam()); }
+
+public:
+  // Tests that value profiles in Record has the same content as (possibly
+  // weighted and sorted) InputVDs for each value kind. ValueProfSorted is true
+  // iff the value profiles of Record are already sorted by count.
+  // InputVDs is a non-const reference since it might need a in-place sort.
+  void testValueDataArray(
+      std::vector<std::pair<uint32_t /* ValueKind */, VDArray &>> &InputVDs,
+      InstrProfRecord &Record, bool ValueProfSorted = false,
+      uint64_t ProfWeight = 1) {
+    for (auto &[ValueKind, InputVD] : InputVDs) {
+      ASSERT_EQ(InputVD.size(), Record.getNumValueSites(ValueKind));
+      for (unsigned i = 0; i < InputVD.size(); i++) {
+        ASSERT_EQ(InputVD[i].size(),
+                  Record.getNumValueDataForSite(ValueKind, i));
+
+        uint64_t WantTotalC = 0, GetTotalC = 0;
+        std::unique_ptr<InstrProfValueData[]> OutputVD =
+            Record.getValueForSite(ValueKind, i, &GetTotalC);
+
+        // If value profile elements of the same instrumented site are sorted by
+        // count in Record, sort the input value data array the same way for
+        // comparison purpose.
+        if (ValueProfSorted) {
+          llvm::stable_sort(InputVD[i], [](const InstrProfValueData &lhs,
+                                           const InstrProfValueData &rhs) {
+            return lhs.Count > rhs.Count;
+          });
+        }
+
+        // The previous ASSERT_EQ already tests the number of value data is
+        // InputVD[i].size(), so there won't be out-of-bound access.
+        for (unsigned j = 0; j < InputVD[i].size(); j++) {
+          EXPECT_EQ(InputVD[i][j].Count * ProfWeight, OutputVD[j].Count);
+          EXPECT_EQ(InputVD[i][j].Value, OutputVD[j].Value);
+          WantTotalC += InputVD[i][j].Count;
+        }
+        EXPECT_EQ(WantTotalC * ProfWeight, GetTotalC);
+      }
+    }
+  }
+
+  // A helper function to test the writes and reads of indirect call value
+  // profiles. The profile writer will scale counters by `ProfWeight` when
+  // adding a record. `Endianness` specifies the endianness used by profile
+  // writer and reader when handling value profile records.
+  void testICallDataReadWrite(
+      uint64_t ProfWeight = 1,
+      llvm::endianness Endianness = llvm::endianness::little) {
```

Comment:
I would like to keep the test assertions/expectations in the test themselves, using this helper makes the tests harder to debug on failure. Here is a slightly different refactoring you can consider: These are already parameterized tests so lets extend the params to include endianness and weights. We can combine the params in INSTANTIATE_TEST_SUITE_P using
```
test::Combine(testing::Bool(), 
  Values(TestParam{/*endian=*/little, /*weight=*/1}, 
               TestParam{/*endian=*/little, /*weight=*/10}, 
               TestParam{/*endian=*/big, /*weight=*/1}
  )
)
```
Here TestParam is a new struct which is introduced. This will keep the existing coverage. You can also increase coverage by combining all cases (as cartesian product).

```
test::Combine(testing::Bool(), // sparse?
    Values(little, big), // endianness
    Values(1, 10), // weights
  )
)
```
What do you think? Feel free to avoid if this is requires more time than you want to invest.

Reply:
At a high level, I'm in favor of keeping test assertions/expectations so that the line numbers are obvious when a test fails.

Future parameterizing the test fixture is a good step, but a cartesian product over all existing test cases of `MaybeSparseInstrProfTest` would create more test cases for all combinations that become vague to reason about (i.e., `value_prof_data_read_write` tests serialization/deserialization of `InstrProfRecord` but invoke profile reader or writer) or  become to adds maintain overhead (i.e., when new aspects are added but not applicable, a test case will do something like `(if std::get<N>(getParam()) == irrelevant) return early`.) 

Since it would probably require more work to refactor the unittest, could I parallel track the test coverage of new feature https://github.com/llvm/llvm-project/pull/66825 and the improvement of unit tests? This way it's less work to keep up with upstream main.

Reply:
Yes, I am also a little wary of what we might find if we do a cartesian product :)
I'm ok to defer the refactoring so that we can account for the new vtable prof test too. Thanks!

Reply:
As chatted offline, I realized test coverage for a new kind of value-type might double the amount of code without factoring out common code once I attempted to add new tests. Thanks for going over the test cases together and discussing about how to properly clean it up and prepare for the new value type. I'm sending out smaller patches.